src/ReservoirPair.sol

```diff
@@ -148,15 +152,47 @@ abstract contract ReservoirPair is IAssetManagedPair, ReservoirERC20 {
         require(aBalance0 <= type(uint104).max && aBalance1 <= type(uint104).max, "RP: OVERFLOW");
         require(aReserve0 <= type(uint104).max && aReserve1 <= type(uint104).max, "RP: OVERFLOW");
 
-        uint32 lBlockTimestamp = uint32(_currentTime());
+        uint32 lBlockTimestamp = _currentTime();
         uint32 lTimeElapsed;
         unchecked {
             // underflow is desired
             // however in the case where no swaps happen in ~68 years (2 ** 31 seconds) the timeElapsed would underflow twice
             lTimeElapsed = lBlockTimestamp - aBlockTimestampLast;
-        }
-        if (lTimeElapsed > 0 && aReserve0 != 0 && aReserve1 != 0) {
-            _updateOracle(aReserve0, aReserve1, lTimeElapsed, lBlockTimestamp);
+
+            // both balance should never be zero, but necessary to check so we don't pass 0 values into arithmetic operations
+            // shortcut to calculate aBalance0 > 0 && aBalance1 > 0
+            if (aBalance0 * aBalance1 > 0) {
+                Observation storage lPrevious = _observations[sSlot0.index];
+                (uint256 lInstantRawPrice, int256 lLogInstantRawPrice) = _calcSpotAndLogPrice(aBalance0, aBalance1);
+
+                // a new sample is not written for the first mint
+                // shortcut to calculate lTimeElapsed > 0 && aReserve0 > 0 && aReserve1 > 0
+                if (lTimeElapsed * aReserve0 * aReserve1 > 0) {
+                    (, int256 lLogInstantClampedPrice) = _calcClampedPrice(
+                        lInstantRawPrice,
+                        lLogInstantRawPrice,
+                        LogCompression.fromLowResLog(lPrevious.logInstantClampedPrice),
+                        lTimeElapsed,
+                        aBlockTimestampLast // assert: aBlockTimestampLast == lPrevious.timestamp
+                    );
+                    _updateOracleNewSample(
+                        sSlot0, lPrevious, lLogInstantRawPrice, lLogInstantClampedPrice, lTimeElapsed, lBlockTimestamp
+                    );
+                } else {
+                    // for instant price updates in the same timestamp, we use the time difference from the previous oracle observation as the time elapsed
+                    lTimeElapsed = lBlockTimestamp - lPrevious.timestamp;
+
+                    (, int256 lLogInstantClampedPrice) = _calcClampedPrice(
+                        lInstantRawPrice,
+                        lLogInstantRawPrice,
+                        LogCompression.fromLowResLog(lPrevious.logInstantClampedPrice),
+                        lTimeElapsed,
+                        lPrevious.timestamp
+                    );
+
+                    _updateOracleInstantPrices(lPrevious, lLogInstantRawPrice, lLogInstantClampedPrice);
```

Comment:
Does this write to the oracle or slot0/cache? Cause with the other values we write them to cache and only update the oracle once per slot. Does this mean that we are writing the oracle every trade, is this fine/a problem - maybe not on L2?

Reply:
this only writes to the oracle observations. This case only happens for 2nd-and-later swaps in the same block.